subprojects/publish/src/main/java/org/gradle/api/publish/internal/PublicationInternal.java

```diff
@@ -74,6 +81,21 @@
 
     boolean isPublishBuildId();
 
+    default void checkForUnpublishableAttributes(SoftwareComponentInternal component, DocumentationRegistry documentationRegistry) {
```

Comment:
I moved this here to avoid duplicating between Ivy and Maven publications, which don't share a common base class.  Perhaps they should, then this could be made non-public?  It's only public on the Internal type though.

Or I could move copies back into `DefaultIvyPublication` and `DefaultMavenPublication`.  Thoughts?

Reply:
I wonder about that. We know that `Configuration` and its hierarchy is going to be a challenge to evolve.
And I am pretty sure there are users out there of `ConfigurationInternal`.

So I would recommend finding a different location for that method, possible as a static one in some helper type.

Reply:
I was thinking that we only needed to block someone from inadvertently publishing the new variants that we've added (sources, jacoco execution data and test results) and not prevent any variant with this category from being published. This is a roundabout way of doing sort of the same thing as https://github.com/gradle/gradle/pull/19056 (which tied publishing to a particular attribute).

Could we just check `isIncubating` in `ConfigurationVariantMapping.collectUsageContexts`?  IIUC, someone would need to call `addVariantsFromConfiguration` with one of the new variant configurations to force it to be added to the software component.